File: logParser.go
package main

import (
	"encoding/csv"
	"flag"
	"fmt"
	"io/ioutil"
	"os"
	"path/filepath"
	"strings"
	"time"
)

func fillMissingParts(parts []string, expectedLength int) []string {
	if len(parts) < expectedLength {
		for i := len(parts); i < expectedLength; i++ {
			parts = append(parts, "")
		}
	}
	return parts
}

func readLines(filePath string) ([]string, error) {
	content, err := ioutil.ReadFile(filePath)
	if err != nil {
		return nil, err
	}
	lines := strings.Split(string(content), "\n")
	return lines, nil
}

func parseLogs(source string, destination string, columnsFile string, verbose bool, separator string) error {
	files, err := filepath.Glob(filepath.Join(source, "*.*"))
	if err != nil {
		return err
	}

	columns, err := readLines(columnsFile)
	if err != nil {
		return err
	}

	var data[][]string

	for i, file := range files {
		if verbose {
			fmt.Printf("Parsing file %d/%d\n", i+1, len(files))
		}

		lines, err := readLines(file)
		if err != nil {
			return err
		}

		for _, line := range lines {
			parts := strings.Split(line, separator)
			parts = fillMissingParts(parts, len(columns))
			data = append(data, parts)
			
		}

		if verbose {
			progressPercentage := (float64(i+1) / float64(len(files))) * 100
			fmt.Printf("Progress: %.2f%%\n", progressPercentage)
		}
	}
	file, err := os.Create(filepath.Join(destination, "results.csv"))
	if err != nil {
		return err
	}
	defer file.Close()

	writer := csv.NewWriter(file)
	defer writer.Flush()

	err = writer.Write(columns)
	if err != nil {
		return err
	}

	err = writer.WriteAll(data)
	if err != nil {
		return err
	}

	fmt.Println("Done!")
	return nil
}

func main() {
	var verbose bool
	var separator string
	flag.BoolVar(&verbose, "v", false, "Show state of files parsing")
	flag.StringVar(&separator, "s", " ", "Specify the separator of the files")
	flag.Parse()

	columnsFile := flag.Arg(0)
	destination := flag.Arg(1)
	source := flag.Arg(2)

	startTime := time.Now()
	err := parseLogs(source, destination, columnsFile, verbose, separator)
	if err != nil {
		fmt.Printf("Error: %v\n", err)
		os.Exit(1)
	}
	endTime := time.Now()

	fmt.Printf("Execution time: %v\n", endTime.Sub(startTime))
}

File: logParser.py
import pandas as pd
import glob
import sys
import argparse
import time


def read_file(file_path: str):
    with open(file_path, 'r', encoding="utf8") as file:
        return file.read().splitlines()
       

def parse_logs(source, destination, columns_file, verbose=False, sep=' '):
    all_files = glob.glob(fr"{source}\*.*")
    columns = read_file(columns_file)
    all_files_len = len(all_files)
   
    data = []
    for i, files in enumerate(all_files, start=1):
        if verbose:
            print(f"Parsing file {i}/{all_files_len}")
            
        file = read_file(files)
        data.extend([line.split(sep)[:len(columns)] for line in file])
            
        if verbose:
            progress_percentage = (i / all_files_len) * 100
            print(f"Progress: {progress_percentage:.2f}%")
        
    df = pd.DataFrame(data, columns=columns)
    print("Saving to csv")
    df.to_csv(f'{destination}\\results.csv', index=False)
    print("Done!")
    

if __name__ == "__main__":
    start_time = time.time()
    parser = argparse.ArgumentParser(description="This tool is designed to parse logfiles having the structure of a txtfile",
                                     formatter_class=argparse.ArgumentDefaultsHelpFormatter,
                                     epilog="Example: python log_parser.py C:\columns.txt C:\destination C:\source -v -s '-' ")
    parser.add_argument("-v", "--verbose", action="store_true", default=False, help="Show state of files parsing")
    parser.add_argument("-s", "--separator", default=' ', help="Specify the separator of the files")
    parser.add_argument("columns", help="Provide the full path of the file containing the columns name. DON'T forget the extension")
    parser.add_argument("destination", help="Give the full path of the folder where the csv will be saved")
    parser.add_argument("source", help="Give the full path of the folder containing the log files")
    args = parser.parse_args()
    parse_logs(args.source, args.destination, args.columns, args.verbose, args.separator)
    end_time = time.time()
    print(f'Execution time: {end_time - start_time} seconds')